### PR TITLE
feat: added vocab_size arg to handle possible mismatch between tokenizer and model

### LINKED ISSUE #19: BUG: Tokenizer length mismatch causes IndexError in NTL initialization (vocab_size < embedding size)
From my observation, NTL seems to initialize certain components based on the tokenizer vocabulary size.  
Currently, the code uses `len(self.tokenizer)` (or equivalently `len(self.tokenizer.get_vocab())`), as shown below:

Image

Image

Image

---

However, this is **inaccurate**.  
I found that in many LLMs/VLMs (at least in my case with Qwen2.5), `len(tokenizer)` is **smaller** than the actual index range used by the model, i.e. `model.get_input_embeddings().weight.shape[0]`, as shown below:

Image

This leads to an error like:

`IndexError: The shape of the mask [151665] at index 0 does not match the shape of the indexed tensor [2, 504, 151936] at index 2`

---

When I was using the Scratch version of the [MWE of NTL](https://github.com/tum-ai/number-token-loss/blob/main/src/ntl/loss_functions/base_number_token_loss.py)
I encountered the same issue.  

My workaround was to add an **explicit function parameter** that takes the true `vocab_size` used by the model during runtime, instead of inferring it from the tokenizer. See below:

Image

Currently, I am using the **PyPI version** (since it supports Wasserstein distance),  
but this issue still needs to be fixed.

---

**Suggestion:**  
Allow `vocab_size` to be explicitly passed into the NTL initialization (e.g., `NumberTokenLoss(vocab_size=model.get_input_embeddings().weight.shape[0], ...)`)  
to ensure compatibility with models where the tokenizer vocabulary length does not match the actual embedding matrix size.

### PR DESCRIPTION
**Pull request by casinca**
This is an attempt to fix #19

- added optional vocab_size argument to NTL
- added a preventive check in `_validate_inputs` to catch mismatches upfront
- fix potential new edge case if numbers tokens are added
- added a test for mismatches

All tests passed

### CODE REVIEW

**Comment by casinca** [comment]
I did this before I saw your comment on the issue.  
Since it's good enough to let the user fix it by himself when he encounters an index error, then I could just simplify: create `self.number_values` tensor with optional `vocab_size` if not None and that's it.

File: ntloss/core.py [hunk 1]
```diff
@@ -34,6 +37,9 @@ def __init__(
         """
         super().__init__()
         self.tokenizer = tokenizer
+        self.user_vocab_size = vocab_size
```
Comment (jannisborn) [inline ntloss/core.py]:
```suggestion
```

File: ntloss/core.py [hunk 2]
```diff
@@ -34,6 +37,9 @@ def __init__(
...
+        self.effective_vocab_size = vocab_size if vocab_size is not None else len(self.tokenizer)
```
[the remark this answers is not in the document]
Comment (jannisborn) [reply ntloss/core.py]: I think this would simplify the logic a bit, also in the error handling downstream

**Review by jannisborn — CHANGES_REQUESTED** [review]
Great job @casinca once again! Please have a look, I think we should just remove the `max` and then it's good to go! I also added a small suggestion on code simplification.

Thanks a ton 🙏🏼

File: ntloss/core.py [hunk 3]
```diff
@@ -51,7 +57,9 @@ def setup_number_tokens(self):
         if vocab_size < len(self.tokenizer) and new_tokens > 0:
             logger.warning(f"Added {new_tokens} new tokens for number token loss")
         vocab = self.tokenizer.get_vocab()
-        self.number_values: FloatTensor = torch.full((len(vocab),), float("nan"))
+
+        final_vocab_size = max(self.effective_vocab_size, len(vocab))
```
[the remark this answers is not in the document]
Comment (casinca) [reply ntloss/core.py]: A potential solution to what I mention above, late initialization of `self.vocab_size` just before `self.number_values` so that we get the exact same behavior as before (when `vocab_size=None`).

```diff
@@ -37,16 +37,15 @@ class AbstractNTLoss(ABC):
         """
         super().__init__()
         self.tokenizer = tokenizer
-        self.vocab_size = vocab_size if vocab_size is not None else len(self.tokenizer)
         self._vocab_size_validated = False
         self.digit_level = digit_level
         self.reweigh = reweigh

-        self.setup_number_tokens()
+        self.setup_number_tokens(vocab_size)

         self.max_dist = torch.tensor(0.0)

-    def setup_number_tokens(self):
+    def setup_number_tokens(self, user_vocab_size: Optional[int] = None):
         """Setting up attributes needed by NT loss"""

         # Add digits to vocab if not there yet.
@@ -56,6 +55,7 @@ class AbstractNTLoss(ABC):
         if vocab_size < len(self.tokenizer) and new_tokens > 0:
             logger.warning(f"Added {new_tokens} new tokens for number token loss")
         vocab = self.tokenizer.get_vocab()
+        self.vocab_size = user_vocab_size if user_vocab_size is not None else len(self.tokenizer)
         self.number_values: FloatTensor = torch.full((self.vocab_size,), float("nan"))

         # Try to convert each token to a float after stripping the space prefix
```

I saw you mentioned you are in vacation so I don't want to annoy you with PR reviews.
Since it's not urgent, we can see this when you get back.

[the remark this answers is not in the document]
Comment (jannisborn) [reply ntloss/core.py]: Thanks for being so considerate @casinca , fortunately I have a lot of fun in this project :)
I see your point that `len(tokenizer)` may differ from `len(tokenizer.get_vocab())` but for standard tokenizers this assumption holds. If not, the user can control it explicitly

**Review by jannisborn — APPROVED** [review]
